tsdb/head.go

```diff
@@ -306,6 +308,64 @@ func newHeadMetrics(h *Head, r prometheus.Registerer) *headMetrics {
 	return m
 }
 
+// HeadStats are the statistics for the head component of the DB.
+type HeadStats struct {
+	WALReplayStatus *WALReplayStatus
+}
+
+// NewHeadStats returns a new HeadStats object.
+func NewHeadStats() *HeadStats {
+	return &HeadStats{
+		WALReplayStatus: &WALReplayStatus{},
+	}
+}
+
+// WALReplayStatus contains status information about the WAL replay.
+type WALReplayStatus struct {
+	sync.RWMutex
+	Read  int
+	Total int
+	// Progess is a progress bar, 0 - 100%.
+	Progress int
+	State    WALReplayState
+}
+
+// WALReplayState is the state of the WAL replay.
+type WALReplayState int
+
+const (
+	// WALReplayWaiting is the state before the WAL replay has started.
+	WALReplayWaiting WALReplayState = iota
+	// WALReplayInProgress is the state while the WAL replay is in progress.
+	WALReplayInProgress
+	// WALReplayDone is the state after the the WAL replay has finished.
+	WALReplayDone
+)
+
+var walReplayStates = []string{
+	"waiting",
+	"in progress",
+	"done",
+}
+
+// String returns the string format of a WALReplayState
+func (state WALReplayState) String() string {
+	if state < WALReplayWaiting || state > WALReplayDone {
+		return "unknown"
+	}
+
+	return walReplayStates[state]
+}
+
+// GetWALReplayStatus returns the WAL replay status information.
+func (s *WALReplayStatus) GetWALReplayStatus() *WALReplayStatus {
```

Comment:
Let's make sure we copy struct (so just return non pointer). We don't want to put pointer as it's still the same pointer to the same WALReplayStatus object so locking does not help ): 
```suggestion
func (s *WALReplayStatus) GetWALReplayStatus() WALReplayStatus {
```

Reply:
Govet complains about copying the lock value here (see failed tests)